keps/sig-api-machinery/NNNN-built-in-default/README.md

```diff
@@ -0,0 +1,302 @@
+building CRDs. CRDs have acquired the `// +default` tag to declaratively
+define defaults and that mechanisms still doesn't exist for built-in.
+This proposal tries to solve that inconsistency.
+
+## Motivation
+
+Having a new `// +default` marker for built-in types has several
+important benefits:
+
+- It's more convenient for types authors.
+- The types and their behaviors are easier to understand since they are
+described together.
+- They are less error-prone, and prevent people from using anti-patterns
+like "cross-field dependency defaults" (default depends on another
+field), non-deterministic or complicated defaults. These patterns
+might be useful in limited circumstances, but it should be easier to do
+the right thing in most cases.
+- Removes the inconsistency with CRDs that already support this
+mechanism.
+- Solves bug where associative keys need to either be required or
+  defaulted.
+
+### Goals
+
+The goal is to add a new `// +default` tag to our current built-in Go
+IDL. That tag will be transformed into the OpenAPI `default` tag and
+then routed to defaulting functions so that defaulting can be done
+declaratively.
+
+### Non-Goals
+
+While this proposal is probably paving the way for additional OpenAPI
+based validation, validation is entirely out-of-scope for this project.
+
+This proposal also doesn't intend to remove the existing defaulting
+mechanism since some defaulting still requires complex logic that can't
+be defined declaratively. We could replace some defaulting that doesn't
+require to be done as code though.
+
+## Proposal
+
+Since the goal of this proposal is to increase consistency with the CRD
+defaulting, we propose to use the same mechanism (both in OpenAPI and
+similar to what kubebuilder does). This defaulting will happen BEFORE
+the existing defaulting functions are called (existing functions can be
+updated to panic so that we can see if defaults are properly applied).
+
+One very important nuance with CRDs is that CRDs are deserialized in
+unstructured objects that can distinguish between specified and
+unspecified fields. That is not the case for built-in types that are
+directly deserialized into Go-struct, where the unspecified fields are
+zero-valued. It's important to understand that only zero-value
+omit-empty fields will be defaulted using that mechanism.
+
+Complex defaulting will not be supported as a declarative marker and
+will continue being supported through the existing defaulting functions.
+
+### Risks and Mitigations
+
+The main risk is a bug that would cause a default to not be set as
+expected, either because the default would be missing or not set as
+expected. This would result in possible conversion failures, invalid
+objects being stored in the database or being served to users.
+
+## Design Details
+
+There are multiple challenges to running defaults declaratively and
+changing the current mechanisms:
+
+- We don't use OpenAPI to drive validation nor defaulting
```

Comment:
Why mention validation? Validation is acting on internal types. That's too late. We have to get into the decoding step, i.e. it must be part of the scheme.